src/cached_object_store.rs

```diff
@@ -762,6 +817,286 @@ impl LocalCacheEntry for FsCacheEntry {
+    tx: tokio::sync::mpsc::Sender<(std::path::PathBuf, usize)>,
+    rx: Mutex<Option<tokio::sync::mpsc::Receiver<(std::path::PathBuf, usize)>>>,
+    evict_task_handle: OnceCell<tokio::task::JoinHandle<()>>,
+    scan_task_handle: OnceCell<tokio::task::JoinHandle<()>>,
+    db_stats: Arc<DbStats>,
+}
+
+impl FsCacheEvictor {
+    pub fn new(
+        root_folder: std::path::PathBuf,
+        max_cache_size_bytes: usize,
+        db_stats: Arc<DbStats>,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::channel(100);
+        Self {
+            root_folder,
+            max_cache_size_bytes,
+            tx,
+            rx: Mutex::new(Some(rx)),
+            evict_task_handle: OnceCell::new(),
+            scan_task_handle: OnceCell::new(),
+            db_stats,
+        }
+    }
+
+    async fn start(&self) {
+        let inner = Arc::new(FsCacheEvictorInner::new(
+            self.root_folder.clone(),
+            self.max_cache_size_bytes,
+            self.db_stats.clone(),
+        ));
+
+        let guard = self.rx.lock();
+        let rx = guard.await.take().expect("evictor already started");
+
+        // scan the cache folder to load the cache entries on disk
+        let scan_task_handle = tokio::spawn(inner.clone().scan_entries(true));
+        self.scan_task_handle.set(scan_task_handle).ok();
+
+        // start the background evictor task, it'll be triggered whenever a new cache entry is added
+        let evict_task_handle = tokio::spawn(Self::background_evict(rx, inner));
+        self.evict_task_handle.set(evict_task_handle).ok();
+    }
+
+    async fn started(&self) -> bool {
+        self.rx.lock().await.is_none()
+    }
+
+    async fn background_evict(
+        mut rx: tokio::sync::mpsc::Receiver<(std::path::PathBuf, usize)>,
+        inner: Arc<FsCacheEvictorInner>,
+    ) {
+        while let Some((path, bytes)) = rx.recv().await {
+            inner
+                .track_new_entry(path, bytes, SystemTime::now(), true)
+                .await;
+        }
+    }
+
+    pub async fn maybe_evict(&self, path: std::path::PathBuf, bytes: usize) {
+        if !self.started().await {
+            return;
+        }
+
+        self.tx.send((path, bytes)).await.ok();
+    }
+}
+
+/// FsCacheEvictorInner manages the cache entries in an in-memory trie, and evict the cache entries
+/// when the cache size exceeds the limit. it uses a pick-of-2 strategy to approximate LRU, and evict
+/// the older file when the cache size exceeds the limit.
+///
+/// On start up, FsCacheEvictorInner will scan the cache folder to load the cache files into the in-memory
+/// trie cache_entries. This loading process is interleaved with the maybe_evict is being called, so the
+/// cache entries should be wrapped with Mutex<_>.
+#[derive(Debug)]
+struct FsCacheEvictorInner {
+    root_folder: std::path::PathBuf,
+    batch_factor: usize,
+    max_cache_size_bytes: usize,
+    cache_entries: Mutex<Trie<std::path::PathBuf, (SystemTime, usize)>>,
```

Comment:
where do we actually update the access times for the cache entries?

Reply:
oh my bad 🥲, yes it's a missed part after changed to memory tracked access time. i'll fix it, thank you for the reminding

Reply:
fixed in b64d08a